src/protocols/bdx/AsyncTransferFacilitator.h

```diff
@@ -0,0 +1,130 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include <lib/core/CHIPError.h>
+#include <lib/support/BitFlags.h>
+#include <messaging/ExchangeContext.h>
+#include <messaging/ExchangeHolder.h>
+#include <protocols/bdx/BdxTransferSession.h>
+#include <transport/raw/MessageHeader.h>
+
+#pragma once
+
+namespace chip {
+namespace bdx {
+
+/**
+ * An abstract class with methods for handling BDX messages received from an ExchangeContext. Once a message is receievd, this
+ * class passes the message to the TransferSession to process the received message and gets the next output events from the
+ * TransferSession state machine and either sends a message accross the exchange or calls the HandleTransferSessionOutput virtual
+ * method to notify the subclass of the event generated. It keeps getting the next output event until it receieves an output event
+ * of type TransferSession::OutputEventType::kNone. For messages that are sent to the HandleTransferSessionOutput method, the
+ * subclass must call the NotifyEventHandled to notify the AsyncTransferFacilitator that the event has been handled and returns an
+ * error code for error cases or success.
+ *
+ * This class does not define any methods for beginning a transfer or initializing the underlying TransferSession object.
+ * See AsyncResponder for a class that does.
+ * TODO: # 29334 - Add AsyncInitiator to handle the initiating side of a transfer.
+ *
+ * An AsyncTransferFacilitator is associated with a specific BDX transfer.
+ */
+class AsyncTransferFacilitator : public Messaging::ExchangeDelegate
+{
+public:
+    AsyncTransferFacilitator() : mExchange(*this) {}
+    ~AsyncTransferFacilitator() override;
+
+    /**
+     * This method should be implemented to contain business-logic handling of BDX messages
+     * and other TransferSession events.
+     *
+     * @param[in] event An OutputEvent that contains the output from the TransferSession object.
+     */
+    virtual void HandleTransferSessionOutput(TransferSession::OutputEvent & event) = 0;
+
+    /**
+     * This method should be implemented to destroy the object subclassing AsyncTransferFacilitator.
+     */
+    virtual void DestroySelf() = 0;
+
+protected:
+    CHIP_ERROR OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                 System::PacketBufferHandle && payload) override;
+    void OnResponseTimeout(Messaging::ExchangeContext * ec) override;
+    void OnExchangeClosing(Messaging::ExchangeContext * ec) override;
+
+    CHIP_ERROR SendMessage(const TransferSession::MessageTypeData msgTypeData, System::PacketBufferHandle & msgBuf);
+
+    static bdx::StatusCode GetBdxStatusCodeFromChipError(CHIP_ERROR err);
+
+    void CleanUp();
```

Comment:
This is protected, right?  So subclasses can call it?  When would they do that?  What is the purpose of this method?

Reply:
This is a private method that calls DestroySelf which should be protected. I wrote this as a utility to release the exchange since we grab the exchange in an exchange holder etc.